gusto/state.py

```diff
@@ -530,3 +534,34 @@ def get_latlon_mesh(mesh):
     op2.par_loop(kernel, coords_latlon.cell_set,
                  coords_latlon.dat(op2.RW, coords_latlon.cell_node_map()))
     return Mesh(coords_latlon)
+
+
+def topo_sort(field_deps):
+    # map node: (input_deps, output_deps)
+    graph = dict((f.name, (list(deps), [])) for f, deps in field_deps)
+    roots = []
+    for f, input_deps in field_deps:
+        if len(input_deps) == 0:
+            # No dependencies, candidate for evaluation
+            roots.append(f)
+        for d in input_deps:
+            # add f as output dependency
+            graph[d][1].append(f)
+
+    schedule = []
+    while roots:
+        n = roots.pop()
+        schedule.append(n)
+        output_deps = list(graph[n.name][1])
+        for m in output_deps:
+            # Remove edge
+            graph[m.name][0].remove(n.name)
+            graph[n.name][1].remove(m)
+            # If m now as no input deps, candidate for evaluation
+            if len(graph[m.name][0]) == 0:
+                roots.append(m)
```

Comment:
Simpler if the scheduler only works with the names (and then you map them back into fields at the end?)

Reply:
possibly... although it is not the field, it is the diagnostic field class. I can get the field name easily from the class (as here) but I can't get the class from the name without defining a dictionary (I don't think) and I couldn't decide what to call the dictionary...

Reply:
I thought something like this:

```python
def topo_sort(field_deps):
    name2field = dict((f.name, f) for f, _ in field_deps)
    # map node: (input_deps, output_deps)
    graph = dict((f.name, (list(deps), [])) for f, deps in field_deps)
    roots = []
    for f, input_deps in field_deps:
        if len(input_deps) == 0:
            # No dependencies, candidate for evaluation
            roots.append(f.name)
        for d in input_deps:
            # add f as output dependency
            graph[d][1].append(f.name)

    schedule = []
    while roots:
        n = roots.pop()
        schedule.append(n)
        output_deps = list(graph[n][1])
        for m in output_deps:
            # Remove edge
            graph[m][0].remove(n)
            graph[n][1].remove(m)
            # If m now has no input deps, candidate for evaluation
            if len(graph[m][0]) == 0:
                roots.append(m)
    if any(len(i) for i, _ in graph.values()):
        cycle = "\n".join("%s -> %s" % (f, i) for f, (i, _) in graph.items()
                          if f not in schedule)
        raise RuntimeError("Field dependencies have a cycle:\n\n%s" % cycle)
    return list(map(name2field.__getitem__, schedule))
```
No?

Reply:
It's marginal, but you don't have to worry about using the `field` when deleting from output deps, but the `name` when indexing the graph, or deleting from input deps.

Reply:
`name2field` is the kind of name I was looking for! I got hung up on the fact that it isn't the field, but the class used to calculate the field - but we don't care about that when we name, for example, the `diagnostic_fields` list.